File: internal/tools/edit.go
package tools

import (
	"fmt"
	"os"
	"strings"
)

func (s *Set) edit(a map[string]any) Result {
	p, _ := a["path"].(string)
	old, ok1 := a["old_string"].(string)
	newText, ok2 := a["new_string"].(string)
	all, _ := a["replace_all"].(bool)
	if p == "" || !ok1 || !ok2 || old == "" {
		return Result{Text: "错误：缺少参数或 old_string 为空"}
	}
	b, err := os.ReadFile(p)
	if err != nil {
		return Result{Text: "错误：" + err.Error()}
	}
	if int64(len(b)) > s.EditMax {
		return Result{Text: "错误：文件超过 2MB，请用 shell 处理"}
	}
	content := string(b)
	n := strings.Count(content, old)
	if n == 0 {
		return Result{Text: "错误：没有找到 old_string"}
	}
	if n > 1 && !all {
		return Result{Text: fmt.Sprintf("错误：old_string 匹配到 %d 处，请增加上下文或设 replace_all=true", n)}
	}
	limit := 1
	if all {
		limit = -1
	}
	result := strings.Replace(content, old, newText, limit)
	if err := os.WriteFile(p, []byte(result), infoMode(p)); err != nil {
		return Result{Text: "错误：" + err.Error()}
	}
	changed := 1
	if all {
		changed = n
	}
	return Result{Text: fmt.Sprintf("已修改 %s（替换了 %d 处）", p, changed)}
}

File: server/agent/functions/edit.js
import fs from "node:fs/promises";
import path from "node:path";

export default async function edit(args, config, signal) {
  if (
    typeof args.path !== "string" ||
    !args.path ||
    typeof args.old_string !== "string" ||
    !args.old_string ||
    typeof args.new_string !== "string"
  ) {
    throw new Error("缺少参数或 old_string 为空");
  }
  const filename = path.resolve(config.workdir || process.cwd(), args.path);
  const info = await fs.stat(filename);
  if (!info.isFile() || info.size > 2 * 1024 * 1024) {
    throw new Error("不是普通文件或超过 2MB");
  }
  const content = await fs.readFile(filename, { encoding: "utf8", signal });
  const count = content.split(args.old_string).length - 1;
  if (!count) {
    throw new Error("没有找到 old_string");
  }
  if (count > 1 && !args.replace_all) {
    throw new Error(`匹配到 ${count} 处，请增加上下文或设置 replace_all`);
  }
  signal?.throwIfAborted();
  const next = args.replace_all
    ? content.split(args.old_string).join(args.new_string)
    : content.replace(args.old_string, () => args.new_string);
  await fs.writeFile(filename, next, { signal });
  return {
    text: `已修改 ${args.path}（替换 ${args.replace_all ? count : 1} 处）`,
  };
}
